### PR TITLE
Use file hash as cache buster

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?
Using the file hash is much more reasonable as cache buster for any file that is served over the web.

### 2. What does this change do, exactly?
Adds a custom less function swhash in js and php which looks up the file in parameter and returns its hash. Current hash algorithm is md5.

### 3. Describe each step to reproduce the issue or behaviour.
1. Replace loading indicator animation file
2. No change happen
3. Image was not reloaded as cache was not busted

### 4. Which documentation changes (if any) need to be made because of this PR?
In the future anyone should use the hash of a file for cache busting

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by JoshuaBehrens** [comment]
I added an app secret that probably is discussable. My thoughts on that: if you have a local copy of a shop running on the same url for reasons you would get the same cached theme files although it is an other instance of the shop. This way the cache busters are valid per installation and do not depend on anything.

I still need to integrate that altered value into the swhash functions.

File: _sql/migrations/1650-add-application-secret-configuration.php [hunk 1]
```diff
@@ -0,0 +1,60 @@
+<?php declare(strict_types=1);
+
+use Shopware\Components\Migrations\AbstractMigration;
+
+class Migrations_Migration1650 extends AbstractMigration
+{
+    public function up($modus)
+    {
+        $secret = bin2hex(random_bytes(32));
```
Comment (shyim) [inline _sql/migrations/1650-add-application-secret-configuration.php]: This is useless, all new installations will have the same one

File: _sql/migrations/1650-add-application-secret-configuration.php [hunk 2]
```diff
@@ -0,0 +1,60 @@
+<?php declare(strict_types=1);
+
+use Shopware\Components\Migrations\AbstractMigration;
+
+class Migrations_Migration1650 extends AbstractMigration
+{
+    public function up($modus)
```
Comment (shyim) [inline _sql/migrations/1650-add-application-secret-configuration.php]: void me

File: _sql/migrations/1650-add-application-secret-configuration.php (hunk 1 above)
Comment (shyim) [reply _sql/migrations/1650-add-application-secret-configuration.php]: I guess you are searching something like this https://github.com/shopware/shopware/blob/60fcfbab6c0397d0317475313672968ece29ef67/engine/Shopware/Bundle/PluginInstallerBundle/Service/UniqueIdGenerator/UniqueIdGenerator.php

File: _sql/migrations/1650-add-application-secret-configuration.php (hunk 2 above)
Comment (JoshuaBehrens) [reply _sql/migrations/1650-add-application-secret-configuration.php]: I just copied the 1649 :sweat_smile: but you are right

File: _sql/migrations/1650-add-application-secret-configuration.php (hunk 1 above)
Comment (JoshuaBehrens) [reply _sql/migrations/1650-add-application-secret-configuration.php]: Can you explain why every installation gets the same random bytes?

Comment (shyim) [reply _sql/migrations/1650-add-application-secret-configuration.php]: The build process executes all migrations and dumps than a install.sql. This install.sql will be used for all installations.

Comment (JoshuaBehrens) [reply _sql/migrations/1650-add-application-secret-configuration.php]: How will the unique id generator help me out there? It looks like I have to generate the id NOT in any migration

Comment (shyim) [reply _sql/migrations/1650-add-application-secret-configuration.php]: Maybe use the uniqueIDGenerator service for your secret. Its generates a unique id for the installation and saves that as config element for later usage.

Comment (JoshuaBehrens) [reply _sql/migrations/1650-add-application-secret-configuration.php]: But when do I generate it? A migration would be a no brainer running once and everything is fine. With that I have to check it every time

Comment (JoshuaBehrens) [reply _sql/migrations/1650-add-application-secret-configuration.php]: Now I get it. It is not about how to generate the key in this migration but use the service as it already provides such an installation specific id. Sorry my mistake. NICE!

**Comment by shyim** [comment]
Hey @JoshuaBehrens,
unfortunately the Grunt implementation does not work.

```
❯ node node_modules/.bin/grunt
Running "fileExists:js" (fileExists) task
Checking 64 files for existence...
OK

Running "less:development" (less) task
>> ../themes/Frontend/Responsive/frontend/_public/src/less/_components/loading-indicator.less: [L92:C37] error evaluating function `swhash`: Cannot read property '@shopware-revision' of null
Warning: Error compiling ../web/cache/all.less Use --force to continue.

Aborted due to warnings.
```

Can you check and fix it?

**Comment by JoshuaBehrens** [comment]
Good finding. Somehow `grunt watch` was not affected. Please check my changes.

**Comment by soebbing** [comment]
Hi @JoshuaBehrens, thank you again for your contribution! 💙

**Comment by mitelg** [comment]
happy 40 accepted PRs anniversary :smiley: :tada: :birthday: :blue_heart:

**Comment by JoshuaBehrens** [comment]
Already 😲❓ insane! Thank you 🙇 You are looking closer to the statistics than I am. I think I'd missed it without you 💙

**Pull request closed without merging**